### DISCUSSION TITLE
How to pass outputs from different stacks?

### DISCUSSION THREAD

**Discussion by fenos** [terramate-io/terramate#525] [Q&A] [answered]
Hello, Thanks for making terramate! I'm evaluating to migrate over from terragrunt

In essence, it is not clear to me how I can reference the outputs from a stack into another stack.
Imagine the following scenario and folder structure:

My infrastructure is divided by `env/region/mod_name` 
I'd like to re-use the output from the `global` region into the others regions

```
stacks
|-- dev
     |--- global
     |       |--- network 
     |       |        |--- import_network.tm.hcl
     |--- europe-west-1
     |      |---- network
     |      |         |--- config.tm.hcl
     |      |         |--- import_subnets.tm.hcl
```

In the `config.tm.hcl` inside the `europe-west-1/network` i'd like to do something like this:

```
globals {
  network_self_link = globals.global.network.output.self_link
}
```

however, I'm not sure how to reference the output, any tips?

**Comment by katcipis** [comment, marked as answer]
Hi @fenos,

Thanks for you interest in Terramate and for taking the time explaining your issue.

Sorry if I misunderstood the question, but if you are asking about getting values from [Terraform outputs](https://www.terraform.io/language/values/outputs) from another stack this is not supported by Terramate in the moment. At least not natively in a seamless way.

Our approach is to use globals as a way to share data accross stacks so you usually don't need the outputs. You can have globals defined on `dev/global/network` and then [import](https://github.com/mineiros-io/terramate/blob/dcc8c364245ff5025513ff203fbdd4778c2a3121/docs/config-overview.md) those globals to use in the stacks, instead of loading them from outputs.

For scenarios where you need something that is only defined at plan/apply time you can define globals that are used to build the resource and then use those same globals and a data source to retrieve whatever you need that only exists after apply. Then for the ordering you can use the [orchestration features](https://github.com/mineiros-io/terramate/blob/dcc8c364245ff5025513ff203fbdd4778c2a3121/docs/orchestration.md), guaranteeing that a stack that creates resource X will be applied before the one that uses a data source to load information about that resource.

In the absence of a specific data source you can also use the [terraform_remote_state data source](https://www.terraform.io/language/state/remote-state-data) to get outputs from another stack, while avoiding any duplications on how to reference the remote state using globals.

Possibly there are other solutions too, just trying to give you some idea of what is possible, hope it helps :smiley:.

**Comment by mariux** [comment]
@fenos in addition to @katcipis answer:

He described the recommended way of sharing results between stacks. You might face the issue that just ordering stacks is not solving all your needs as you need to apply stack-A before stack-B can be even planned.

You can use a neat Terraform behavior to get around this. When you add a dependency to a [`null_resource`](https://registry.terraform.io/providers/hashicorp/null/latest/docs/resources/resource) on the data sources used, terraform will postpone reading the data source to he apply phase and plan with "(known after apply)" values. 

This is working fine for initial deployments but can get more tricky and complex over time.

**Comment by az-z** [comment]
why not to use Terra Cloud workspaces for output?
